pkg/cmd/run.go

```diff
@@ -74,67 +102,111 @@ doppler run --token=123 -- printenv`,
+	fetchSecrets := !(enableFallback && fallbackOnly)
+	if !fetchSecrets {
 		return readFallbackFile(fallbackPath)
 	}
 
-	response, err := http.GetSecrets(localConfig.APIHost.Value, utils.GetBool(localConfig.VerifyTLS.Value, true), localConfig.Token.Value, localConfig.EnclaveProject.Value, localConfig.EnclaveConfig.Value)
-	if err != (http.Error{}) {
-		if useFallbackFile {
+	response, httpErr := http.DownloadSecrets(localConfig.APIHost.Value, utils.GetBool(localConfig.VerifyTLS.Value, true), localConfig.Token.Value, localConfig.EnclaveProject.Value, localConfig.EnclaveConfig.Value, true)
+	if httpErr != (http.Error{}) {
+		if enableFallback {
+			utils.LogDebug("Failed to fetch secrets from the API")
 			return readFallbackFile(fallbackPath)
 		}
-		utils.HandleError(err.Unwrap(), err.Message)
+		utils.HandleError(httpErr.Unwrap(), httpErr.Message)
 	}
 
-	if useFallbackFile && !fallbackReadonly {
-		err := ioutil.WriteFile(fallbackPath, response, 0600)
-		if err != nil {
-			utils.HandleError(err, "Unable to write fallback file")
+	// ensure the response can be parsed before proceeding
+	secrets, err := parseSecrets(response)
+	if err != nil {
+		if enableFallback {
+			utils.LogDebug("Failed to parse the API response")
+			return readFallbackFile(fallbackPath)
 		}
+		utils.HandleError(err, "Unable to parse API response")
 	}
 
-	secrets, parseErr := models.ParseSecrets(response)
-	if parseErr != nil {
-		utils.HandleError(parseErr, "Unable to parse API response")
+	writeFallbackFile := enableFallback && !fallbackReadonly
+	if writeFallbackFile {
+		utils.LogDebug(fmt.Sprintf("Writing to fallback file %s", fallbackPath))
+		err := ioutil.WriteFile(fallbackPath, response, 0600)
+		if err != nil {
+			utils.LogDebug("Failed to write to fallback file")
+			if exitOnWriteFailure {
+				utils.HandleError(err, "Unable to write fallback file", strings.Join(writeFailureMessage(), "\n"))
+			} else {
+				utils.LogDebug("Not exiting due to --no-exit-on-write-failure flag")
+			}
+		}
 	}
 
-	secretsStrings := map[string]string{}
-	for key, value := range secrets {
-		secretsStrings[key] = value.ComputedValue
-	}
+	return secrets
+}
 
-	return secretsStrings
+func writeFailureMessage() []string {
+	var msg []string
+
+	msg = append(msg, "")
+	msg = append(msg, color.Green.Render("Why did doppler exit?"))
+	msg = append(msg, "Doppler failed to make a local backup of your secrets, known as a fallback file.")
+	msg = append(msg, "The most common cause for this is insufficient permissions, including using a fallback file created by a different user.")
+	msg = append(msg, "")
+	msg = append(msg, color.Green.Render("Why does that matter?"))
+	msg = append(msg, "Without the fallback file, your secrets would be inaccessible in the event of a network partition or Doppler downtime.")
```

Comment:
`Without the fallback file, your secrets would be inaccessible in the event you do not have internet access or Doppler downtime.`

Reply:
I much prefer the current messaging

Reply:
Renamed `network partition` to `network outage`